### PR TITLE
CDC-ACM usb serial reliability improvements

### PR DESCRIPTION
**Pull request by bigdinotech**
-add delay between transmissions
-check DTR status before transmitting
-only toggle TX led only when serial port is open

### CODE REVIEW

**Comment by bigdinotech** [comment]
@SidLeung @eriknyquist please review

File: src/cdcacm_serial.c [hunk 1]
```diff
@@ -95,19 +95,25 @@ static void write_data(struct device *dev, const char *buf, int len)
...
 				Tx_TAIL = (Tx_TAIL + 1)% SBS;
 			}
-			write_data(dev, (const char*)write_buffer, cnt);
-			gpio_pin_write(gpio_dev, TXRX_LED, 1);	//turn TXRX led off
+			uart_line_ctrl_get(dev, LINE_CTRL_DTR, &dtr);
+			if(dtr)
+			{
+				gpio_pin_write(gpio_dev, TXRX_LED, 0);	//turn TXRX led on
+				write_data(dev, (const char*)write_buffer, cnt);
+				k_busy_wait(40 * cnt);
```
Comment (eriknyquist) [inline src/cdcacm_serial.c]: Is this usecs? msecs? or ticks? can you give it a name and specify, i.e. DELAY_TICKS or DELAY_USECS or whetever it is

**Comment by bigdinotech** [comment]
@eriknyquist gave the delay value a define. Please review

File: src/cdcacm_serial.c [hunk 2]
```diff
@@ -95,19 +96,25 @@ static void write_data(struct device *dev, const char *buf, int len)
...
-			gpio_pin_write(gpio_dev, TXRX_LED, 0);	//turn TXRX led on
+			int cnt = 0;			
 			for (; (Tx_TAIL != Tx_HEAD) && (cnt < BUFFER_LENGTH*2);cnt++)
 			{
 				write_buffer[cnt] = Tx_BUFF[Tx_TAIL];
 				Tx_TAIL = (Tx_TAIL + 1)% SBS;
 			}
-			write_data(dev, (const char*)write_buffer, cnt);
-			gpio_pin_write(gpio_dev, TXRX_LED, 1);	//turn TXRX led off
+			uart_line_ctrl_get(dev, LINE_CTRL_DTR, &dtr);
```
Comment (eriknyquist) [inline src/cdcacm_serial.c]: Why reading the DTR here? what does it tell you?

trying to figure it out myself....
DTR normally used for flow control right? i.e. receiver on the other end will assert DTR when it's ready to receive more data. So it should be asserted, most of the time.

So, basically you are only doing the delay when DTR is asserted. Shouldn't you do it for every character?

Comment (eriknyquist) [inline src/cdcacm_serial.c]: I just don't get why you need to check DTR....

Comment (bigdinotech) [inline src/cdcacm_serial.c]: When the serial port is opened dtr = 1
When the serial port is closed dtr = 0

**Pull request merged**